extract.py

```diff
@@ -0,0 +1,408 @@
+
+    time = (d[0].header['expstart'] + d[0].header['expend'])/(2.0) + 2400000.5                    #converts time to BJD_TDB; see Eastman et al. 2010 equation 4
+    time = time + (32.184)/(24.0*60.0*60.0)
+    time = time + (suntimecorr.suntimecorr(ancil.ra, ancil.dec, array([time]), ancil.coordtable[ancil.visnum], verbose=False))/(60.0*60.0*24.0)
+
+    phase = (time-ancil.t0)/ancil.period - math.floor((time-ancil.t0)/ancil.period)
+    if phase > 0.5: phase = phase - 1.0                                    #calculates orbital phase
+
+    template_waves = ancil.wave_grid[0, int(ancil.refpix[0,1]) + ancil.LTV1, cmin:cmax]             #LK interpolation 8/18
+    shift = 0.
+    #corrects for wavelength drift over time
+    if convert_to_bool(obs_par['correct_wave_shift']) == True:
+        if nspectra == 0:
+            shift = 0.
+            template_spectrum = spec_opt                                #makes the first exposure the template spectrum
+            best_spec = spec_opt
+            best_var = var_opt
+        else:
+            #shifts spectrum so it matches the template
+            [best_spec, best_var, shift] = interpolate_spectrum(spec_opt, np.sqrt(var_opt), template_spectrum, template_waves)
+            best_var = best_var**2
+
+            spec_opt = best_spec                                    #saves the interpolated spectrum
+            var_opt = best_var
+
+    #if convert_to_bool(obs_par['plot_spectrum']) == True: plot_spectrum(ancil)
+    #print sum(spec_opt)/sum(var_opt), sum(spec_box)/sum(var_box), sum(spec_opt)/sum(spec_box)
+
+    """plt.plot(template_waves, cRate)
+    plt.plot(template_waves, cRate1)
+    plt.plot(template_waves, cRate2)
+    plt.show()"""
+
+    #plt.plot(template_waves, spec_opt)
+    #plt.show()
+
+    n = len(spec_opt)
+    print(phase[0], sum(spec_opt), sum(var_opt),  sum(spec_box), sum(var_box), time[0], ancil.visnum, ancil.orbnum, scan)
+    if ancil.output == True:
+        print(phase[0], sum(spec_opt), sum(var_opt),  sum(spec_box), sum(var_box), time[0], ancil.visnum, ancil.orbnum, scan, np.mean(cRate), file=whitefile)
+        for ii in arange(n):
+            print(time[0], phase[0], spec_opt[ii], var_opt[ii], template_waves[ii], ancil.visnum, ancil.orbnum, scan, cRate[ii], file=specfile)
+        print(nspectra, time[0], numoutliers, skymedian, shift, file=diagnosticsfile)
+
+    nspectra += 1
+    if nspectra%10 == 0: print("Extraction", '{0:1f}'.format(float(nspectra)/float(len(ancil.files))*100.), "% complete, time elapsed (min) =", '{0:0.1f}'.format(clock()/60.))
+
+    #ind = np.isnan(spec_opt)
+    #plt.plot(template_waves, spec_opt)
+    #plt.plot(template_waves, ind*6e5)
+    #plt.show()
+
+    print("length of spectrum", len(spec_opt))
+    print("# nans", sum(np.isnan(spec_opt)))
+    print(nspectra, time[0], sum(spec_opt), np.sum(spec_box), ancil.visnum, f, shift)
+
+if ancil.output == True:
```

Comment:
Outputs are better done all together and files should be closed at the end of the appropriate loop.